### DISCUSSION TITLE
Ngspice s2p - issue with Ngspice43+

### DISCUSSION THREAD

**Discussion by iwbnwif** [ra3xdh/qucs_s#1036] [Q&A] [answered]
There appears to be an issue with s2p simulation and the recent Ngspice 43+. I have tried the  s2p simulation from here: [issues/233#issuecomment-1571140202 ](https://github.com/ra3xdh/qucs_s/issues/233#issuecomment-1571140202) and get the error: h

```
Error on line 14 or its substitute:
  a.xsub1.x1.a_gen_11 %vd xsub1.x1.10 0 %vd xsub1.x1.11 xsub1.x1.12 xsub1.x1:gen_model_11
Null not allowed for parameter (null) on model xsub1.x1:gen_model_11.
```
Ngspice42 seems to work normally. I downloaded and built Ngspice43+ from the tarball on the Ngspice home page.

This is on Ubuntu 24.04 using the current Qucs-S appimage.

**Comment by ra3xdh** [comment]
I have Ngspice-43 on Arch(Manjaro) and cannot reproduce this error. The simulation works as expected. Please check the build options of the new Ngspice package. Something may be disabled by default. Ngspice also may be referencing the XSPICE libraries from 42 package.

**Comment by iwbnwif** [comment]
Thank you for looking into this. I tried rebuilding with --with-x and --enable-cider as well as using the "compile_linux_dist.sh" script and the result is the same in all cases. 

I will try to debug the issue further if it is not normally reproducible.

Sorry, but is the use of s2p files documented somewhere? I thought I had read it in the readthedocs pages but can no longer find it. The example I referenced above is not straightforward as it doesn't seem to use the s2p file directly.

**Comment by ra3xdh** [comment, marked as answer]
> I tried rebuilding with --with-x and --enable-cider as well as using the "compile_linux_dist.sh" 

These options have no effect on this. This feature is implemented via XSPICE. You may also check the spiceinit contents. Maybe Ngspice loads wrong XSPICE libraries. 

> Sorry, but is the use of s2p files documented somewhere? 

The information from readthedocs is totally outdated and even misleading. Never use this documentation source. Unfortunately the usage of SnP files and many other new features are not documented. Writing a documentation requires much more time than writing a code. The only source about S2P files support is this thread: https://github.com/ra3xdh/qucs_s/issues/233 The S2P file is converted to Ngspice frequency defined source that is implemented as XSPICE library.

**Comment by iwbnwif** [comment]
> You may also check the spiceinit contents. Maybe Ngspice loads wrong XSPICE libraries.

Thank you, after much recompiling and searching it finally turns out that this indeed was the problem! But in fact it was the `spinit` file rather than `spiceinit` on my system.

For anyone having similar problems, in particular building on Debian / Ubuntu you might like to check out the configure parameters that are used by the package maintainers. You can find the build log under https://launchpad.net/ubuntu/+source/ngspice. It will be under the Ubuntu version (e.g., Oracular) and then under Builds <arch>. Search the build log for <configure> and the information should be there.

The next thing to consider is if there are any other versions of `ngspice` on your system. I also had `Kicad` installed which had the libs and some files but not the executable. 

It is not enough to run the built file as the executable looks for some other files, including the `spinit` mentioned above. This behavior can be adapted through environment variables (see the Ngspice manual) but by default it is looking in system folders. Probably, the easiest method is to `sudo make install` which puts the required files in the right places (assuming you used the packager's configure options). Caveat is making sure that you don't have files from other installations hanging around. In general apt purge is your friend here ;)

**Comment by tomhajjar** [comment]
I'm using ngspice 43+

Attached are two s-parameter test projects. Data is included so you can compare results. 

S-parameter_files_ngspice_prj has some data files created to diagnose some problems we were having in the beginning.

[S-parameters_ngspice.zip](https://github.com/user-attachments/files/17608608/S-parameters_ngspice.zip)

**Reply by iwbnwif** [reply]
Thank you, I will take a look. At the moment it is working quite well and I am getting the expected results (as far as my knowledge allows!)
